# Implementing logistic regression from scratch

The goal of this notebook is to implement your own logistic regression classifier. You will:

 * Extract features from Amazon product reviews.
 * Convert an pandas DataFrame into a NumPy array.
 * Implement the link function for logistic regression.
 * Write a function to compute the derivative of the log likelihood function with respect to a single coefficient.
 * Implement gradient ascent.
 * Given a set of coefficients, predict sentiments.
 * Compute classification accuracy for the logistic regression model.
 
Let's get started!

In [ ]:
# Import some libs

import pandas
import numpy as np
from sklearn.model_selection import train_test_split

### Load review dataset
For this assignment, we will use a subset of the Amazon product review dataset. The subset was chosen to contain similar numbers of positive and negative reviews, as the original dataset consisted primarily of positive reviews.

Load the dataset into a data frame named **products**. One column of this dataset is **sentiment**, corresponding to the class label with +1 indicating a review with positive sentiment and -1 for negative sentiment.

Let us quickly explore more of this dataset. The **name** column indicates the name of the product. Try listing the name of the first 10 products in the dataset.

After that, try counting the number of positive and negative reviews.

In [ ]:
# Import amazon_baby.csv data to pandas dataframe
products_df = pandas.read_csv('/content/drive/MyDrive/FUNIX Progress/MLP303x_1.1-A_EN/data/amazon_baby_subset.csv')

In [ ]:
products_df["name"][:10]

0    Stop Pacifier Sucking without tears with Thumb...
1      Nature's Lullabies Second Year Sticker Calendar
2      Nature's Lullabies Second Year Sticker Calendar
3                          Lamaze Peekaboo, I Love You
4    SoftPlay Peek-A-Boo Where's Elmo A Children's ...
5                            Our Baby Girl Memory Book
6    Hunnt&reg; Falling Flowers and Birds Kids Nurs...
7    Blessed By Pope Benedict XVI Divine Mercy Full...
8    Cloth Diaper Pins Stainless Steel Traditional ...
9    Cloth Diaper Pins Stainless Steel Traditional ...
Name: name, dtype: object

In [ ]:
products_df.groupby('sentiment').count()

name  review  rating
sentiment                       
-1         26461   26393   26493
 1         26521   26438   26579

### Apply text cleaning on the review data
In this section, we will perform some simple feature cleaning using data frames. The last assignment used all words in building bag-of-words features, but here we limit ourselves to 193 words (for simplicity). We compiled a list of 193 most frequent words into the JSON file named **important_words.json**. Load the words into a list **important_words**.

In [ ]:
import json

with open('/content/drive/MyDrive/FUNIX Progress/MLP303x_1.1-A_EN/data/important_words.json', 'r') as f:
    important_words = json.loads(f.read())
print(important_words)

['baby', 'one', 'great', 'love', 'use', 'would', 'like', 'easy', 'little', 'seat', 'old', 'well', 'get', 'also', 'really', 'son', 'time', 'bought', 'product', 'good', 'daughter', 'much', 'loves', 'stroller', 'put', 'months', 'car', 'still', 'back', 'used', 'recommend', 'first', 'even', 'perfect', 'nice', 'bag', 'two', 'using', 'got', 'fit', 'around', 'diaper', 'enough', 'month', 'price', 'go', 'could', 'soft', 'since', 'buy', 'room', 'works', 'made', 'child', 'keep', 'size', 'small', 'need', 'year', 'big', 'make', 'take', 'easily', 'think', 'crib', 'clean', 'way', 'quality', 'thing', 'better', 'without', 'set', 'new', 'every', 'cute', 'best', 'bottles', 'work', 'purchased', 'right', 'lot', 'side', 'happy', 'comfortable', 'toy', 'able', 'kids', 'bit', 'night', 'long', 'fits', 'see', 'us', 'another', 'play', 'day', 'money', 'monitor', 'tried', 'thought', 'never', 'item', 'hard', 'plastic', 'however', 'disappointed', 'reviews', 'something', 'going', 'pump', 'bottle', 'cup', 'waste', 'retu

Here we remove **remove_punctuation** and fill in N/A's with empty reivew

In [ ]:
def remove_punctuation(text):
    import string
    return text.translate(text.maketrans('', '', string.punctuation))

products_df = products_df.fillna({'review':''})  # fill in N/A's in the review column
products_df['review_clean'] = products_df['review'].apply(remove_punctuation)

Now we proceed with the second item. For each word in **important_words**, we compute a count for the number of times the word occurs in the review. We will store this count in a separate column (one for each word). The result of this feature processing is a single column for each word in **important_words** which keeps a count of the number of times the respective word occurs in the review text.

In [ ]:
# 1-hot encoding
for word in important_words:
    products_df[word] = products_df['review_clean'].apply(lambda s : s.split().count(word))

products_df.head()

name  ... either
0  Stop Pacifier Sucking without tears with Thumb...  ...      0
1    Nature's Lullabies Second Year Sticker Calendar  ...      0
2    Nature's Lullabies Second Year Sticker Calendar  ...      0
3                        Lamaze Peekaboo, I Love You  ...      0
4  SoftPlay Peek-A-Boo Where's Elmo A Children's ...  ...      0

[5 rows x 198 columns]

Now, write some code to compute the number of product reviews that contain the word perfect.
<br>
**Quiz Question:** How many reviews contain the word **"perfect"**?
<br>
**Your answer:** 

In [ ]:
len(products_df[products_df['perfect'] != 0])

2955

### Convert data frame to multi-dimensional array
Write a function that extracts columns from a data frame and converts them into a multi-dimensional array. We plan to use them throughout the course, so make sure to get this function right.
<br>
The function should accept three parameters:
<br>
- **dataframe:** a data frame to be converted
- **features:** a list of string, containing the names of the columns that are used as features.
- **label:** a string, containing the name of the single column that is used as class labels.
<br>
<br>
The function should return two values:
<br>
- one 2D array for features
- one 1D array for class labels
<br>
<br>
The function should do the following:
<br>
- Prepend a new column constant to dataframe and fill it with 1's. This column takes account of the intercept term. Make sure that the constant column appears first in the data frame.
- Prepend a string 'constant' to the list features. Make sure the string 'constant' appears first in the list.
- Extract columns in dataframe whose names appear in the list features.
- Convert the extracted columns into a 2D array using a function in the data frame library. If you are using Pandas, you would use as_matrix() function.
- Extract the single column in dataframe whose name corresponds to the string label.
- Convert the column into a 1D array.
- Return the 2D array and the 1D array.

In [ ]:
def get_numpy_data(dataframe, features, label):
    dataframe['constant'] = 1

    features = ['constant'] + features
    
    feature_matrix = dataframe[features].values
    label_array = dataframe[label].values

    return (feature_matrix, label_array)

Using the function **get_numpy_data**, extract two arrays **feature_matrix** and **sentiment**. The 2D array **feature_matrix** would contain the content of the columns given by the list **important_words**. The 1D array sentiment would contain the content of the column **sentiment**.

In [ ]:
feature_matrix, sentiment = get_numpy_data(products_df, important_words, 'sentiment')

**Quiz Question:** How many features are there in the feature_matrix?
<br>
**Your answer:**

In [ ]:
feature_matrix.shape

(53072, 194)

Recall from lecture that the link function is given by
<br>
$P(y_i = +1 | \mathbf{x}_i, \mathbf{w}) = \dfrac{1}{1 + \exp{(-\mathbf{w}^\intercal h(\mathbf{x}_i))}}$
<br>
where the feature vector $h(\mathbf{x}_i)$ represents the word counts of **important_words** in the review $\mathbf{x}_i$
<br>
Write a function named **predict_probability** that implements the link function.
<br>
- Take two parameters: **feature_matrix** and **coefficients**.
- First compute the dot product of **feature_matrix** and **coefficients**.
- Then compute the link function $P(y = +1 | \mathbf{x}, \mathbf{w})$
- Return the predictions given by the link function.

In [ ]:
def predict_probability(feature_matrix, coefficients):
    # Take dot product of feature_matrix and coefficients
    scores = feature_matrix.dot(coefficients)
    
    # Compute P(y_i = +1 | x_i, w) using the link function
    predictions = 1 / (1 + np.exp(-scores))
    
    # return predictions
    return scores, predictions

**Aside**. How the link function works with matrix algebra

Since the word counts are stored as columns in **feature_matrix**, each i-th row of the matrix corresponds to the feature vector $h(\mathbf{x}_i)$:
$$[\mbox{feature_matrix}] = \left[\begin{array}{c} h(\mathbf{x}_1)^\intercal \\ h(\mathbf{x}_2)^\intercal \\ \vdots \\ h(\mathbf{x}_N)^\intercal\end{array}\right] = \left[\begin{array}{cccc}h_0(\mathbf{x}_1) & h_1(\mathbf{x}_1) & \cdots & h_D(\mathbf{x}_1) \\ h_0(\mathbf{x}_2) & h_1(\mathbf{x}_2) & \cdots & h_D(\mathbf{x}_2) \\ \vdots & \vdots & \ddots & \vdots \\ h_0(\mathbf{x}_N) & h_1(\mathbf{x}_N) & \cdots & h_D(\mathbf{x}_N) \end{array}\right]$$
By the rules of matrix multiplication, the score vector containing elements $\mathbf{w}^\intercal h(\mathbf{x}_i)$ is obtained by multiplying **feature_matrix** and the coefficient vector $\mathbf{w}$:
$$[\mbox{score}] = [\mbox{feature_matrix}]\mathbf{w} = \left[\begin{array}{c} h(\mathbf{x}_1)^\intercal \\ h(\mathbf{x}_2)^\intercal \\ \vdots \\ h(\mathbf{x}_N)^\intercal\end{array}\right] \mathbf{w} = \left[\begin{array}{c} h(\mathbf{x}_1)^\intercal \mathbf{w} \\ h(\mathbf{x}_2)^\intercal \mathbf{w} \\ \vdots \\ h(\mathbf{x}_N)^\intercal \mathbf{w}\end{array}\right] = \left[\begin{array}{c} \mathbf{w}^\intercal h(\mathbf{x}_1) \\ \mathbf{w}^\intercal h(\mathbf{x}_2) \\ \vdots \\ \mathbf{w}^\intercal h(\mathbf{x}_N) \end{array}\right]$$

### Compute derivative of log likelihood with respect to a single coefficient
Recall from lecture:
$$\displaystyle \frac{\partial \ell}{\partial w_j} = \sum_{i=1}^N h_j(\mathbf{x}_i) (\mathbf{1}[y_i = +1] - P(y_i = +1 | \mathbf{x}_i, \mathbf{w}))$$
- **errors:** vector whose i-th value contains
$$\mathbf{1}[y_i = +1] - P(y_i = +1 | \mathbf{x}_i, \mathbf{w})$$
- **feature:** vector whose i-th value contains
$h_j(\mathbf{x}_i)$
<br>
This corresponds to the j-th column of feature_matrix.
<br>
The function should do the following:
<br>
- Take two parameters **errors** and **feature**.
- Compute the dot product of **errors** and **feature**.
- Return the dot product. This is the derivative with respect to a single coefficient w_j.

In [ ]:
def feature_derivative(errors, feature):     
    # Compute the dot product of errors and feature
    derivative = np.dot(np.transpose(errors), feature)

    return derivative

In the main lecture, our focus was on the likelihood. In the advanced optional video, however, we introduced a transformation of this likelihood---called the log-likelihood---that simplifies the derivation of the gradient and is more numerically stable. Due to its numerical stability, we will use the log-likelihood instead of the likelihood to assess the algorithm.

The log-likelihood is computed using the following formula (see the advanced optional video if you are curious about the derivation of this equation):
$$\displaystyle \ell \ell (\mathbf{w}) = \sum_{i=1}^N \Big( (\mathbf{1}[y_i = +1] - 1) \mathbf{w}^\intercal h(\mathbf{x}_i) - \ln{\big(1 + \exp{(-\mathbf{w}^\intercal h(\mathbf{x}_i) )} \big)} \Big)$$
<br>
Write a function **compute_log_likelihood** that implements the equation
<br>
The function has two parameters:
<br>
**indicator**: Has shape (N, 1). **indicator[i]** = True if **yi** = +1 otherwise **indicator[i]** = False
<br>
**scores**: The scores return by the **predict_probability** function. Refer to the above section for more details about the **scores** parameter.

In [ ]:
def compute_log_likelihood(indicator, scores):
    lp = np.sum((indicator-1)*scores - np.log(1. + np.exp(-scores)))
    return lp

### Taking gradient steps
Now we are ready to implement our own logistic regression. All we have to do is to write a gradient ascent function that takes gradient steps towards the optimum.
<br>
Write a function **logistic_regression** to fit a logistic regression model using gradient ascent.
<br>
The function accepts the following parameters:
<br>
- **feature_matrix:** 2D array of features
- **sentiment:** 1D array of class labels
- **initial_coefficients:** 1D array containing initial values of coefficients
- **step_size:** a parameter controlling the size of the gradient steps
- **max_iter:** number of iterations to run gradient ascent
<br>
The function returns the last set of coefficients after performing gradient ascent.
<br>
The function carries out the following steps:
<br>
1. Initialize vector **coefficients** to **initial_coefficients**.
2. Predict the class probability $P(y_i = +1 | \mathbf{x}_i,\mathbf{w})$ using your **predict_probability** function and save it to variable **predictions**.
3. Compute indicator value for $(y_i = +1)$ by comparing **sentiment** against +1. Save it to variable **indicator**.
4. Compute the errors as difference between **indicator** and **predictions**. Save the errors to variable **errors**.
5. For each j-th coefficient, compute the per-coefficient derivative by calling **feature_derivative** with the j-th column of **feature_matrix**. Then increment the j-th coefficient by (step_size*derivative).
6. Once in a while, insert code to print out the log likelihood.
7. Repeat steps 2-6 for **max_iter** times.

In [ ]:
from math import sqrt
def logistic_regression(feature_matrix, sentiment, initial_coefficients, step_size, max_iter):
    coefficients = initial_coefficients # make sure it's a numpy array

    for itr in range(max_iter):
        # Predict P(y_i = +1|x_1,w) using your predict_probability() function
        scores, predictions = predict_probability(feature_matrix, coefficients)

        # Compute indicator value for (y_i = +1)
        indicator = (sentiment == +1) # .reshape(predictions.shape[0], 1)

        # Compute the errors as indicator - predictions
        errors = indicator - predictions

        for j in range(coefficients.shape[0]): # loop over each coefficient
            # Recall that feature_matrix[:,j] is the feature column associated with coefficients[j]
            # compute the derivative for coefficients[j]. Save it in a variable called derivative
            derivative = feature_derivative(errors, feature_matrix[:,j])

            # add the step size times the derivative to the current coefficient
            coefficients[j] += step_size * derivative
        
        if itr % 100 == 0:
            lp = compute_log_likelihood(indicator, scores)
            print ('iteration: {}, log likelihood: {}'.format(itr, lp))

    return coefficients

In [ ]:
# Advance
def self_make_logistic_regression(feature_matrix, sentiment, initial_coefficients, step_size, max_iter):
    coefficients = initial_coefficients

    for itr in range(max_iter):
        scores, predictions = predict_probability(feature_matrix, coefficients)

        indicator = (sentiment == +1)

        errors = indicator - predictions
        
        # Modify here
        coefficients += step_size * feature_derivative(errors, feature_matrix)
        
        if itr % 100 == 0:
            lp = compute_log_likelihood(indicator, scores)
            print ('iteration: {}, log likelihood: {}'.format(itr, lp))

    return coefficients

Now, let us run the logistic regression solver with the parameters below:
<br>
- **feature_matrix** = feature_matrix extracted
- **sentiment** = sentiment extracted
- **initial_coefficients** = a 194-dimensional vector filled with zeros
- **step_size** = 1e-7
- **max_ite**r = 301
<br>
Save the returned **coefficients** to variable **coefficients**.

In [ ]:
initial_coefficients = np.zeros((feature_matrix.shape[1],))
step_size = 1e-7
max_iter = 301
coefficients = logistic_regression(feature_matrix, sentiment, initial_coefficients, step_size, max_iter)

iteration: 0, log likelihood: -36786.70716667741
iteration: 100, log likelihood: -36235.70757728472
iteration: 200, log likelihood: -35733.69577367936
iteration: 300, log likelihood: -35272.93069887046


In [ ]:
self_make_coefficients = self_make_logistic_regression(feature_matrix, sentiment, initial_coefficients, step_size, max_iter)

iteration: 0, log likelihood: -35268.51212682766
iteration: 100, log likelihood: -34844.003998221226
iteration: 200, log likelihood: -34451.20771660703
iteration: 300, log likelihood: -34086.653432478925


In [ ]:
coefficients - self_make_coefficients

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0.])

**Quiz question:** As each iteration of gradient ascent passes, does the log likelihood increase or decrease?
<br>
**Your answer: Increase**

### Predicting sentiments
Recall from lecture that class predictions for a data point x can be computed from the coefficients w using the following formula:
<br>
$$\hat{y}_i = \begin{cases} +1 & \text{if }\mathbf{x}_i^\intercal \mathbf{w} > 0 \\ -1 & \text{if }\mathbf{x}_i^\intercal \mathbf{w} \leq 0\end{cases}$$
Now, we write some code to compute class predictions. We do this in two steps:
- First compute the scores using feature_matrix and coefficients using a dot product.
- Then apply threshold 0 on the scores to compute the class predictions. Refer to the formula above.
<br>

**Quiz question:** How many reviews were predicted to have positive sentiment?
<br>
**Your answer:**

In [ ]:
# scores = feature_matrix.dot(coefficients)
scores, predictions = predict_probability(feature_matrix, coefficients)

In [ ]:
len(predictions[predictions > 0.5])

26452

### Measuring accuracy
We will now measure the classification accuracy of the model. Recall from the lecture that the classification accuracy can be computed as follows:
$$\mbox{accuracy} = \dfrac{\mbox{# correctly classified data points}}{\mbox{# total data points}}$$
**Quiz question:** hat is the accuracy of the model on predictions made above? (round to 2 digits of accuracy)
<br>
**Your answer: 0.75**

In [ ]:
diff = (scores > 0) == (sentiment > 0)
print ('Accuracy: {}'.format(np.sum(diff) / len(diff)))

Accuracy: 0.7548047934880916


Recall that in the earlier assignment, we were able to compute the "**most positive words**". These are words that correspond most strongly with positive reviews. In order to do this, we will first do the following:
- Treat each coefficient as a tuple, i.e. (**word, coefficient_value**). The intercept has no corresponding word, so throw it out.
- Sort all the (**word, coefficient_value**) tuples by **coefficient_value** in descending order. Save the sorted list of tuples to **word_coefficient_tuples**.

Now, **word_coefficient_tuples** contains a sorted list of (**word, coefficient_value**) tuples. The first 10 elements in this list correspond to the words that are most positive.
<br>
**Quiz question:** What is the top 10 positive words?
<br>
**Your answer:**
<br>
**Quiz question:** What is the top 10 negative words?
<br>
**Your answer:**

In [ ]:
coefficients = list(coefficients[1:]) # exclude intercept
word_coefficient_tuples = [(word, coefficient) for word, coefficient in zip(important_words, coefficients)]
word_coefficient_tuples = sorted(word_coefficient_tuples, key=lambda x:x[1], reverse=True)

In [ ]:
word_coefficient_tuples[:11]

[('great', 0.12617483799732268),
 ('love', 0.125265596682835),
 ('easy', 0.1232886335285767),
 ('loves', 0.08677892963900478),
 ('little', 0.08648965018920975),
 ('well', 0.05821798832990093),
 ('perfect', 0.05785064299368017),
 ('old', 0.03806010368745643),
 ('nice', 0.03589639699960394),
 ('daughter', 0.034298130985858315),
 ('soft', 0.03395785092036801)]

In [ ]:
word_coefficient_tuples[-10:]

[('waste', -0.04624412301305021),
 ('back', -0.050620242563483915),
 ('get', -0.050983258735577776),
 ('return', -0.05103450496244153),
 ('even', -0.055659458248917736),
 ('disappointed', -0.05632740024143928),
 ('work', -0.0617400094008335),
 ('money', -0.07426071913860344),
 ('product', -0.07453419871769011),
 ('would', -0.09446749481057254)]